### ISSUE TITLE
Should IReadableChannel interface really have two awaitables?

### ISSUE THREAD

**Issue by TimLovellSmith** [davidfowl/Channels#76]
Re:
[IReadableChnanel.cs @ 6cd3090615e5a8250dc04c814fbc10eabb3e7bfa](https://github.com/davidfowl/Channels/blob/6cd3090615e5a8250dc04c814fbc10eabb3e7bfa/src/Channels/IReadableChannel.cs)

`Task Completion { get;  }` "Gets a task that completes when the channel is completed and has no more data to be read. "

Just wondering, is Task really a very useful way to signal completion? It _suggests_ to me upon reading the API either that a) I should have create two separate suspended Tasks in memory that are in a race with each other, one waiting for Read, one waiting for EOF, and that one or both of these tasks will know about how to clean up the other - or that b) I will have to use a single logical thread listening to both events by way of await Task.WhenAny() etc.

Either of these approaches to using the API seem complicated to me. 

(I compare with familiar e.g. Read() apis which return empty byte array for end of stream, the IEnumerable<T> API which has HasNext(), or IObservable<T> which works via func callbacks for OnNext/OnError/OnCompleted, so it can all easily be done with one thread of execution.)

Also documentation of the interface contract was unclear to me - I what is implied about whether Read() and Completion can both become signalled (unblock) at the same time? Is it possible? Or is there really a rule that e.g. Completion awaitable can only becomes unblocked after all bytes have been read from the channel?

**Comment by TimLovellSmith** [comment]
I guess TLDR would be: reading this interface to me it's quite unclear when is the best time to start awaiting Completed.

**Comment by TimLovellSmith** [comment]
PS, I looked at ChannelHttpConent and it is using Completion task as if it were a bool..., not a Task... so serves as an illustration of how I feel this API is maybe overly complex compared to really having a bool IsCompleted? https://github.com/davidfowl/Channels/blob/master/samples/Channels.Samples/HttpClient/ChannelHttpContent.cs

**Comment by davidfowl** [comment]
@TimLovellSmith You never await the `Completion` Task. It's being used as a tri-state to mean the channel is over, did it complete with an error or successfully. It could be a separate type but. Task as an API has an advantage that it can be awaited, registered with a callback, or polled for status.

The API is unclear though and leads to loops like this:

https://github.com/davidfowl/Channels/blob/master/src/Channels/ChannelExtensions.cs#L108-L134

**Comment by TimLovellSmith** [comment]
@davidfowl Hmm. I see... In that example, you are check Completion.IsCompleted, but I think that's a bug: if the channel is faulted, that just returns true and the exception never bubbles up!

it would be more correct to do if (inputBuffer.IsEmpty && fin) 
{
    var dummy = inputBuffer.Completion; // to raise any exception in the Completion task
    return;
}

Right?

**Comment by TimLovellSmith** [comment]
Not that I am trying to endorse the Task model... :grin:

**Comment by davidfowl** [comment]
> @davidfowl Hmm. I see... In that example, you are check Completion.IsCompleted, but I think that's a bug: if the channel is faulted, that just returns true and the exception never bubbles up!

The call to ReadAsync throws if the channel ended with in failure.

**Comment by TimLovellSmith** [comment]
@davidfowl So I think you don't need a tristate Completion then: ReadAsync() handles all the exception cases. (And IsCompleted can return false when ReadAsync is in the terminal throwing state.)

**Comment by davidfowl** [comment]
> @davidfowl So I think you don't need a tristate Completion then: ReadAsync() handles all the exception cases. (And IsCompleted can return false when ReadAsync is in the terminal throwing state.)

You missed this one thing as well:

> Task as an API has an advantage that it can be awaited, registered with a callback, or polled for status.

**Comment by KrzysztofCwalina** [comment]
I found it very useful to have a way to get the buffer in a synchronous (read "fast") way, even if the buffer is empty. Then "Fetch" is asynchronous and can be called if the buffer returned from Get is too small. 
This way I can implement opportunistic logic, which is often faster, e.g:

var buffer = channel.Get();
if(buffer.Length < needed) await channel.Fetch(needed - buffer.Length);
var writtenBytes = Write(buffer);
channel.Commit(writtenBytes);

**Comment by benaadams** [comment]
Maybe is should be

``` csharp
if (inputBuffer.IsEmpty && fin)
{
    await input.Completion;
    return;
}
```

to throw cancellation/exception if anything was refactored to behave differently?

**Comment by TimLovellSmith** [comment]
> You missed this one thing as well:
> 
> > Task as an API has an advantage that it can be awaited, registered with a callback, or polled for status.

bool IsCompleted is already pollable for status, so nix that one...

Are the other two really added features?

You do have a way to await IsCompleted or register a callback: do it on Read.

```
{
    try { await Read(); } catch (Exception e) { ... stream error?? }
    if (IsCompleted) { .. stream gracefully closed?? }
    // why didn't we actually read the buffer too? Is some other thread doing that? Do we want to race with them? Or maybe this is the read thread but we just don't care to read data that is on an errored stream because we know we won't be able to respond etc?
}
```

Also I'm just curious: do we think there is anybody who cares about whether a stream is closed/errored even though nobody is reading data from it?

**Comment by davidfowl** [comment]
We'll need to reconcile this as the current pattern is too error prone

**Comment by davidfowl** [comment]
But consider the following example:

https://github.com/davidfowl/Channels/blob/master/samples/Channels.Samples/Framing/Codec.cs#L48

If I don't see an entire frame, I don't always want to consume everything. I want to wait until there's more data to appear before I proceed with consuming bytes. In that case the buffer isn't empty as the channel is queuing data and pushing it back to the consumer. If you know there's no more data coming and you haven't seen a full frame then you need to be able to check that. That's why it's not exactly like stream's `Read()=0`

**Comment by benaadams** [comment]
I think it currently makes more sense then when it was `Completion`

**Comment by halter73** [comment]
It's nice to allow the producer to tell consumer to continue (i.e. ReadAsync() returns) without any data, but I agree the return value of ReadAsync(), ReadableBuffer, should indicate whether or not the end of input is reached.

I think this could be done with an IsFin property on ReadableBuffer. If IsFin is true, IsEmpty must also be true to make the consuming code simpler. This would be the equivalent to a zero length read on a Stream. If IsFin is false, and IsEmpty is true this would indicate the producer is is telling the consumer to continue (possibly to check some out-of-band state) without any new data or closing the Channel.

Checking a the IsCompleted property on Channel is not ideal, because the consumer has to call ReadAsync() after it sees IsCompleted is true.

**Comment by halter73** [comment]
> If IsFin is true, IsEmpty must also be true to make the consuming code simpler

This actually doesn't make sense, because the consumer might not have completely consumed the previous ReadableBuffer.

**Comment by benaadams** [comment]
> I think this could be done with an IsFin property on ReadableBuffer. 

The buffer doesn't necessarily have anything to do with a channel. It could come from a PreservedBuffer for example.

Also `private int _length;` and `private Memory<byte> _first;` are performance cached values; if they go then the `ReadableBuffer` becomes two `ReadCursor`s which are a ptr + int (8 bytes + 4 bytes); which together make 32 bytes (padding); which with [First Class Structs](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/first-class-structs.md) makes `ReadableBuffer` a zero cost register value. 

Adding an extra variable reduces that scope (though the `ReadCursor`s could be unwrapped and .ctor'd on the fly to give an extra 8 bytes of storage on x64)

**Comment by benaadams** [comment]
The channel could throw if
- When the last `ReadAsync` completed it was `IsCompleted == true`
- No `Advance` was called or Advance was called of length 0
- `ReadAsync` is called again

As it would mean there would be no extra data to come; so the await would be futile?

**Comment by benaadams** [comment]
Might be slightly wrong on First Class Structs seems to only go to 8 bytes unless a SIMD type.

Raised question on registers https://github.com/dotnet/coreclr/issues/7525

**Comment by halter73** [comment]
Fixed by #125?

**Comment by davidfowl** [comment]
Resolved
